classes/local/opensesame_handler.php

```diff
@@ -0,0 +1,517 @@
+        $coursedata->summaryformat = FORMAT_HTML;
+        if (!empty($oscourse->descriptionhtml)) {
+            $coursedata->summary = $oscourse->descriptionhtml;
+        }
+        $coursedata->summary = $oscourse->descriptiontext;
+        $coursedata->summary .= "<br>Publisher Name: {$oscourse->publishername}<br>Duration: {$oscourse->duration}";
+        $coursedata->category = $this->extract_category_id_from_os_string(
+            $oscourse->oscategories
+        );
+        $courseid = $DB->get_field('course', 'id', ['idnumber' => $coursedata->idnumber]);
+        if (empty($courseid)) {
+            $course = create_course($coursedata);
+            $courseid = $course->id;
+        } else {
+            $coursedata->id = $courseid;
+            update_course($coursedata);
+        }
+        $oscourse->courseid = $courseid;
+        return '';
+    }
+
+
+    /**
+     * Processes open sesame course entity.
+     * @param opensesame_course $oscourse
+     * @param opensesame $api
+     * @return string Error message or empty
+     * @throws \file_exception
+     */
+    public function process_created_to_imageimported(opensesame_course &$oscourse, opensesame $api): string {
+        $thumbnailurl = $oscourse->thumbnailurl;
+        $courseid = $oscourse->courseid;
+        $context = context_course::instance($courseid);
+        $fileinfo = [
+            'contextid' => $context->id,    // ID of the context.
+            'component' => 'course',        // Your component name.
+            'filearea'  => 'overviewfiles', // Usually = table name.
+            'itemid'    => 0,               // Usually = ID of row in table.
+            'filepath'  => '/',             // Any path beginning and ending in /.
+            'filename'  => 'courseimage_' . $courseid . '.jpg',   // Any filename.
+        ];
+        // Create course image.
+        $fs = get_file_storage();
+        // Make sure there is not an image file to prevent an image file conflict.
+        $fs->delete_area_files($context->id, 'course', 'overviewfiles', 0);
+        // Create a new file containing the text 'hello world'.
+        $fs->create_file_from_url($fileinfo, $thumbnailurl);
+        return '';
+    }
+
+    /**
+     * Processes open sesame course entity.
+     * @param opensesame_course $oscourse
+     * @param opensesame $api
+     * @return string Error message or empty
+     */
+    public function process_imageimported_to_scormimported(opensesame_course &$oscourse, opensesame $api): string {
+        $courseid = $oscourse->courseid;
+        $allowedtype = get_config('tool_opensesame', 'allowedtypes');
+
+        if ($allowedtype == SCORM_TYPE_LOCAL) {
+            $message = $this->get_os_scorm_package($oscourse->packagedownloadurl, $courseid, $api);
+        } else { // AICC type.
+            // Ensure that Admin settings are set to support AICC Launch Urls.
+            $config = new auto_config();
+            $config->configure();
```

Comment:
Does this need to be called for every course? It seems that this would only need to be called one time if AICC courses are included in the download. If we are downloading 10,000 AICC courses, this would be a lot of wasted time writing the config settings to the database for each course.

Reply:
This is no longer need it, about the retried, It is scheduled 5 minutes after the task fails, We can scheduled to retry every 24h maybe? What do you think?

Reply:
My question has nothing to do with retries. The question is whether we need to call configure for every course. It is not a per-course setting so setting it every time we import an AICC course seems very wasteful.

Reply:
Hi, sorry I mixed with the other question, you are right, this config is no longer need it, shouldn't impact the performance at all maybe 0.01s or something like that, but I deleted it anyway.

Reply:
Doesn't it need to be called at least once?   Just not 10,000 times.  Is there another place it can be called from where we would only need to call it one time instead of once per course?

Reply:
@ojnadjarm I think you've addressed everything except this last question.

Reply:
Hi actually yes, but It's done manually here so, it no need to be done anymore. 
![image](https://github.com/OpenSesame/moodle-tool_opensesame/assets/146741983/f9d6e408-1f2f-49b1-b118-9fa51f1b572e)